weasyprint/layout/pages.py

```diff
@@ -453,8 +454,19 @@ def page_width(box, context, containing_block_width):
 def page_height(box, context, containing_block_height):
     page_width_or_height(VerticalBox(context, box), containing_block_height)
 
-
-def make_page(context, root_box, page_type, resume_at, content_empty):
+def box_text_contents(box):
+    if isinstance(box, boxes.TextBox):
+        return box.text
+    elif isinstance(box, boxes.ParentBox):
+        return ''.join(box_text_contents(child) for child in box.children)
```

Comment:
Although that is not the common case, this code will create lots of intermediate strings if the tree is deep. Please this line with:

```
return ''.join(b.text for b in box.descendants() if isinstance(b, boxes.TextBox))
```
(